invoices/invoices.go

```diff
@@ -471,6 +471,20 @@ func (i *Invoice) HTLCSetCompliment(setID *[32]byte,
 	return htlcSet
 }
 
+// IsKeysend returns true if the invoice is a Keysend invoice.
+func (i *Invoice) IsKeysend() bool {
+	// TODO(positiveblue): look for a more reliable way to tests if
+	// an invoice is keysend.
+	return len(i.PaymentRequest) == 0 && !i.IsAMP()
+}
+
+// IsAMP returns true if the invoice is an AMP invoice.
+func (i *Invoice) IsAMP() bool {
+	return i.Terms.Features.HasFeature(
+		lnwire.AMPOptional,
```

Comment:
Also, if we touch this, shouldn't we also add ` || HasFeature(lnwire.AMPRequried)` to future proof it? Or will it never be possible to make an invoice AMP required?

Reply:
I had the same doubt. I wasn't sure when an invoice would have `AMPOptional` and not `AMPRequired`. It is not only that we set the `AMPRequired` bit in our invoices but we never put the `AMPOptional` one.

The trick here is that `HasFeature` checks that `either the bit is set directly OR the queried bit has the same meaning as its corresponding even/odd bit, which is set instead.`

Long history short if it was confusing for me when I was writing the function and for you when you were reviewing it we should change it to `HasFeature(lnwire.AMPRequried)` direactly :ok_hand:

Reply:
Ah I see. Good to know.